### PR TITLE
feat(bridge-ui): use fork timestamp for protocol version detection

### PR DESCRIPTION
**Pull request by ggonzalez94**
## Summary

Make the bridge fork aware so it can be deployed before the fork timestamp

---

> [!NOTE]
> **Medium Risk**
> Changes core bridge proof/sync decision logic and introduces runtime protocol detection that depends on RPC reads and time comparisons, which could affect claimability if mis-detected or if routing config is incomplete.
> 
> **Overview**
> Bridge UI is now *fork-aware*: it detects whether a route should use Pacaya or Shasta by reading `shastaForkTimestamp` from the destination `SignalService` and comparing it to the chain timestamp (with short-lived caching), allowing deployments before the fork without config toggles.
> 
> Proof generation and processability checks are updated to branch by protocol: Pacaya uses `getSyncedChainData(STATE_ROOT, 0)` (via a new `pacayaSignalServiceAbi` kept for backward compatibility) and supports optional Pacaya *multi-hop* proof generation from new `hops` route config, while Shasta continues using Anchor / `CheckpointSaved` and performs checkpoint verification + preflight `verifySignalReceived` only on Shasta.
> 
> Config/types are extended to label Shasta vs Pacaya addresses, add `HopAddressConfig`/`hops`, and tests are adjusted to cover Pacaya multi-hop behavior and client-null handling.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit ef5c687680ead803b08c7d62937d9778265385e4. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: packages/bridge-ui/src/libs/protocol/protocolVersion.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    const forkTimestamp = await readContract(config, {
+      address: signalService,
+      abi: signalServiceForkRouterAbi,
+      functionName: 'shastaForkTimestamp',
+      chainId: destChainId,
+    });
+
+    // Compare current time with fork timestamp
+    const now = BigInt(Math.floor(Date.now() / 1000));
+    const version = now < forkTimestamp ? ProtocolVersion.PACAYA : ProtocolVersion.SHASTA;
```
[the remark this answers is not in the document]
Comment (ggonzalez94) [reply packages/bridge-ui/src/libs/protocol/protocolVersion.ts]: Fixed. `getProtocolVersion` now compares `shastaForkTimestamp` against destination chain time (latest block timestamp) instead of local wall clock. See `packages/bridge-ui/src/libs/protocol/protocolVersion.ts`.

File: packages/bridge-ui/src/libs/protocol/protocolVersion.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+    cache.set(cacheKey, { version, expiry: Date.now() + CACHE_TTL_MS });
+    return version;
+  } catch {
+    // SignalService not upgraded to fork router yet = still Pacaya
+    cache.set(cacheKey, { version: ProtocolVersion.PACAYA, expiry: Date.now() + CACHE_TTL_MS });
+    return ProtocolVersion.PACAYA;
```
[the remark this answers is not in the document]
Comment (ggonzalez94) [reply packages/bridge-ui/src/libs/protocol/protocolVersion.ts]: Fixed. On RPC/read errors we now return PACAYA without caching to avoid pinning the wrong fork; cache expiry is also bounded to the fork time when PACAYA. See `packages/bridge-ui/src/libs/protocol/protocolVersion.ts`.

File: packages/bridge-ui/src/libs/proof/BridgeProver.ts [hunk 3]
```diff
@@ -61,298 +61,310 @@ export class BridgeProver {
...
+      if (!block?.hash || block.number === null) {
+        throw new BlockNotFoundError({ blockNumber });
       }
-    } catch {
-      throw new BlockNotFoundError({ blockNumber: latestSyncedBlock });
-    }
 
-    // Verify state root from L1 RPC matches checkpoint on L2
-    if ((block.stateRoot as Hex) !== checkpointStateRoot) {
-      log('STATE ROOT MISMATCH between L1 RPC and L2 checkpoint', {
```
[the remark this answers is not in the document]
Comment (ggonzalez94) [reply packages/bridge-ui/src/libs/proof/BridgeProver.ts:199]: Fixed. `generateMultiHopProof` now queries blocks and `eth_getProof` on the source chain (`previousChainId`) for each hop, preventing wrong-chain roots/proofs. Added unit coverage in `BridgeProver.test.ts`.

File: packages/bridge-ui/src/libs/proof/BridgeProver.ts [hunk 4]
```diff
@@ -368,142 +380,94 @@ export class BridgeProver {
...
+        data: encodePacked(
+          ['bytes4', 'uint64', 'address', 'bytes32', 'bytes'],
+          [
+            '0x14e72d1d', // verifySignalReceived selector
+            BigInt(srcChainId),
+            bridgeAddress,
+            msgHash,
+            encodedProof,
+          ],
+        ),
```
[the remark this answers is not in the document]
Comment (ggonzalez94) [reply packages/bridge-ui/src/libs/proof/BridgeProver.ts]: Fixed. Pre-flight `verifySignalReceived` now uses `encodeFunctionData` with `signalServiceAbi` (ABI encoding), not `encodePacked`. See `BridgeProver.ts`.

File: packages/bridge-ui/src/libs/proof/BridgeProver.ts [hunk 5]
```diff
@@ -61,298 +61,310 @@ export class BridgeProver {
...
+      routingContractsMap[Number(srcChainId)][Number(destChainId)].bridgeAddress,
+      msgHash,
+    );
+
+    const hopProofs: HopProof[] = [];
+    let previousChainId = srcChainId;
+
+    for (let i = 0; i < hops.length; i++) {
+      const hop = hops[i];
+      const signalServiceAddress = i + 1 < hops.length ? hops[i + 1].signalServiceAddress : destSignalService;
```
[the remark this answers is not in the document]
Comment (ggonzalez94) [reply packages/bridge-ui/src/libs/proof/BridgeProver.ts]: Fixed. Hop proofs now resolve the signal service address from the source-chain mapping for the current hop (fallback to hop config) so the address matches the chain being queried. Added a unit test asserting the address used in `eth_getProof`.

File: packages/bridge-ui/src/libs/protocol/protocolVersion.ts [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+      functionName: 'shastaForkTimestamp',
+      chainId: destChainId,
+    });
+
+    // Compare current time with fork timestamp
+    const now = BigInt(Math.floor(Date.now() / 1000));
+    const version = now < forkTimestamp ? ProtocolVersion.PACAYA : ProtocolVersion.SHASTA;
+
+    cache.set(cacheKey, { version, expiry: Date.now() + CACHE_TTL_MS });
+    return version;
```
[the remark this answers is not in the document]
Comment (ggonzalez94) [reply packages/bridge-ui/src/libs/protocol/protocolVersion.ts:74]: Fixed. Cache expiry is now capped at the fork time based on chain timestamp, so PACAYA isn’t cached past the fork window. See `protocolVersion.ts`.

File: packages/bridge-ui/src/libs/bridge/isTransactionProcessable.ts [hunk 7]
```diff
@@ -31,69 +32,80 @@ const anchorGetBlockStateAbi = [
...
+    address: destSignalService,
+    abi: signalServiceAbi,
+    eventName: 'CheckpointSaved',
+    fromBlock,
+    toBlock: currentBlock,
+  });
+
+  if (logs.length === 0) return null;
+  return BigInt(logs[logs.length - 1].args.blockNumber!);
+}
```
[the remark this answers is not in the document]
Comment (ggonzalez94) [reply packages/bridge-ui/src/libs/bridge/isTransactionProcessable.ts:111]: Agree this is duplicated. I didn’t refactor here to avoid behavior changes between `BridgeProver` and `isTransactionProcessable` (error handling differs). Suggest a focused follow-up to centralize the helper.

File: packages/bridge-ui/src/libs/bridge/isTransactionProcessable.ts [hunk 8]
```diff
@@ -1,16 +1,17 @@
...
 import { isL2Chain } from '$libs/chain';
+import { getProtocolVersion, ProtocolVersion } from '$libs/protocol/protocolVersion';
 import { getLogger } from '$libs/util/logger';
 import { config } from '$libs/wagmi';
 
 import { type BridgeTransaction, MessageStatus } from './types';
 
 const log = getLogger('libs:bridge:isTransactionProcessable');
 
-// ~33 hours on L1 with 12s blocks
```
[the remark this answers is not in the document]
Comment (ggonzalez94) [reply packages/bridge-ui/src/libs/bridge/isTransactionProcessable.ts:13]: Agree on the duplication; deferring to a follow-up PR to move the constant/ABI into shared modules to keep this change scoped to functional fixes.

**Comment by ggonzalez94** [comment]
@codex please review again

**Review by davidtaikocha — APPROVED** [review]
